### ISSUE TITLE
Std Main Loop

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#3075]
Each platform comes with its own “main event loop” implementation, making it hard to build crossplatform API over it.

We shoul provide a base abstract loop definition that can help this.

Discussion opened on the topic.

**Comment by georgkoester** [comment]
I am working on integrating Haxe code in a program that has its own main event loop. I would just like to have that remain possible - nme makes this really hard and I wouldn't want Haxe to go into the same direction.

**Comment by ncannasse** [comment]
Ok, I'm starting to push some ideas around this.
Here's a first concept of haxe abstract "MainLoop":

https://gist.github.com/ncannasse/8674575efe456cdca5ed

It's very basic, just a list of pending events with priority.
My idea is to have the Haxe compiler automatically insert a MainLoop.run() (if MainLoop.hx is referenced and not DCE) after the Haxe main(). 

Systems that want to redefine the MainLoop would only have to respect the API so users can insert their own events in a crossplatform manner.

This could be slightly improved to add a delay parameter to events so we can do proper scheduling without having to run "infinite" loops.

Would like to get @hughsando @underscorediscovery @jgranick @RobDangerous feedback on this, so we can generalize it for all windowing systems (nme, lime, snow, kha)

With this we could even implement Socket events and Timers on platforms that don't have an actual loop such as Neko.

**Comment by underscorediscovery** [comment]
My gut feeling is that it is difficult to make sense of without being heavy handed or invasive. 

In your given example: having no control over what goes inside `while(true)` is just not an option for many use cases. I wonder too about callbacks on C++ being Dynamic calls and not being optimal yet - this is overhead on _every tick_ now we're forced upon. 

If the user would call MainLoop.tick() and not run() once, this would work already because platform differences. There are many other issues like exclusivity and control.

It's important to consider platforms like iOS and Web run the main loop for rendering applications entirely differently, there is no while true and there can't really be a loop like this. Mobile has heavily platform specific ways that tie into the application life cycle and removing control over these aspects for the end user to facilitate this is quite impractical. A Haxe application (if you choose to build your code as one) is just an entry point and not _an application_ which I think is an important separation – the entry points, loops, and more are dictated by the platform and not the programming language - otherwise you are no longer targeting languages but runtimes.

For things (like Elliot mentions) such as Coroutines (of which there is an implementation which just requires a scheduler/event loop), Promises, and other language level constructs that would be great to add to Haxe - and things that require updates like Timers and Sockets - The _concept_ would be a welcome/required addition - however I think it needs care to not be too heavy handed or invasive if it even is possible.

My feeling is that the focus and attention should be on facilitating language constructs and services like sockets - where the implementation can be handled in a _target_ specific way rather than trying to dictate the application life cycle for a bunch of targets, platforms, and use cases that Haxe itself as a programming language has no concern with.

**Comment by ncannasse** [comment]
@underscorediscovery thanks for the feedback.

Regarding "every tick" overweight : we're talking here about a few closure calls per "frame", I don't think that's an overweight. The goal is not for users to register hundreds of events that get called every frame. If they have too many timers for instance, we could plug a single "timers handle" event that would then process all the timers.

Regarding iOS or exotic runtime loops : I understand that sometimes things can be different, but this simple MainLoop is abstract enough to be implemented differently whatever the platform. If you want for instance to just register an Android activity and return immediately, it's entirely possible. The only guarantee you have to give is to call "often" the events that the user has registered.

Regarding if this should be a common construct for cross platform: we have been hearing quite often from our users that it is difficult to extend existing frameworks with reusable code. I think that it's the role of Haxe to provide such abstractions when possible on which frameworks can be built on. Again we are not imposing a particular implementation, only the API/specification.

**Comment by delahee** [comment]
Dynamic access is not much a problem if it doesn't generate alloc and the
code you shown seems devoid of it, so fine the overhead is nil.

As to the overall problem, I think it' is a very good move to propose a
main loop specification. This will help the ecosystem to have entry point
proliferation and that is a very good thing.

I do hope we'll see more "main loop" libraries and even a "shallow
hxcpp/neko" loop and stuff like that that will ease user to create their
frameworks upon basic loops rather than embed huge frameworks and use 5% of
it.

So...I don't have true feedbacks except, go on this is a good thing to me.







2016-03-05 16:38 GMT+01:00 Nicolas Cannasse <notifications@github.com>:

> @underscorediscovery <https://github.com/underscorediscovery> thanks for
> the feedback.
>
> Regarding "every tick" overweight : we're talking here about a few closure
> calls per "frame", I don't think that's an overweight. The goal is not for
> users to register hundreds of events that get called every frame. If they
> have too many timers for instance, we could plug a single "timers handle"
> event that would then process all the timers.
>
> Regarding iOS or exotic runtime loops : I understand that sometimes things
> can be different, but this simple MainLoop is abstract enough to be
> implemented differently whatever the platform. If you want for instance to
> just register an Android activity and return immediately, it's entirely
> possible. The only guarantee you have to give is to call "often" the events
> that the user has registered.
>
> Regarding if this should be a common construct for cross platform: we have
> been hearing quite often from our users that it is difficult to extend
> existing frameworks with reusable code. I think that it's the role of Haxe
> to provide such abstractions when possible on which frameworks can be built
> on. Again we are not imposing a particular implementation, only the
> API/specification.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/issues/3075#issuecomment-192674279>
> .
>



-- 
David Elahee

**Comment by underscorediscovery** [comment]
I don't understand how "often" can translate to "main loop in every framework" then. 
I also don't understand the usage pattern from the user (framework) perspective either because the code shown is not compatible - so if for instance on js will you run `setTimeout(...)` instead of `while(true)` – This can't be used to run a rendering main loop. So what is the framework supposed to use it for? 

As mentioned, I think the idea of an event loop is a good one, but the idea of trying to abstract this away from the user doesn't align as described. If the construct is supposed to be initiated by the user (and not implicitly by Haxe) then it doesn't answer important questions like sockets and coroutines and promises needing an implicit event loop - which makes me curious if the solution is solving a problem that is being constructed to fit multiple situations that don't overlap.

The supposed guarantee just doesn't match requirements for my mind. I also don't consider a few conditions and null checks light weight for the % of use case described in the very hottest path code. If I measure by % of the hot path vs % of described uses - you say only a few - then I can't not ask why 100% of the hot path code is affected with this overhead that wasn't there before.

**Comment by nadako** [comment]
Wouldn't it be sufficient to extract the body inside that `while(true)` in a separate `tick` function and call it in different ways depending on platform (e.g. setTimeout/setImmediate on JS, `while(true)` on some custom platform, a os-specific callback on mobiles, etc)?

**Comment by RobDangerous** [comment]
I don't see any performance problems here, it's not a hot path just because it's in the main loop. I agree with @delahee - as long as it doesn't force allocations on us it's fine.
Only problem I see: "often" is not a very precise specification, I would propose to at least add per frame callbacks. Kha, for comparison, just provides one callback that's called per frame and can optionally run a complete Scheduler on top of that.

PS: Here are the proper instructions for main loops on iOS: http://kode.tech/snippet-1-own-the-main-loop-in-ios/

**Comment by ncannasse** [comment]
@underscorediscovery In JS case I guess the run would simply do something such as:

```
static function run() {
#if js
    requestAnimationFrame(function() {
      tick(); // process events
      run();
   });
#else
   while( true ) {
      ...
   }
#end 
}
```

Coroutines, promises, sockets etc : every high level async feature needs to boil down somehow to an event loop in the end, they can be built on top of this. We don't have any actual way to implement cross platform asynchronous things in Haxe and that's what I'm proposing to solve here. We can think afterwards at how we can implement higher level things.

@RobDangerous yes I agree with the one-time-per-frame spec in case there is some display occurring.

**Comment by back2dos** [comment]
I cannot say much about this as I am yet to find a polite way to say what I think of both the design and the implementation of the current proposal. Maybe if `MainLoop.add(function () {}); MainLoop.run();` didn't [enter an infinite loop](http://try.haxe.org/#17bB1) (caution: running the sample will make your browser hang) I'd find it easier to verbalize my actual concerns. I don't mean to be harsh, I just find it genuinely alarming that everyone seems to have formed an opinion without even running the code. That is no way to arrive at a practical solution, let alone at one that would be worth becoming a standard.

As far as autowrapping is concerned, you can accomplish that with macros. It requires `addGlobalMetaData` to register a build macro which skips all classes but the `-main` class which can be determined by parsing `Sys.args()`.

There is no reason to integrate this with the compiler, so I propose to develop it as a haxelib and discuss standardisation once the code was actually used in the field under a wide spectrum of use cases.

**Comment by RobDangerous** [comment]
Hey @back2dos the way I understand it this discussion is not about the implementation because that would eventually be provided/overridden by the framework depending on the needs of the target system.

**Comment by underscorediscovery** [comment]
@ncannasse you asked specifically about generalizing it for windowing. I gave my opinion on the matter and I consider it problematic in practice.

**Comment by underscorediscovery** [comment]
@ncannasse to further clarify, I don't actually see what it is that is being suggested because the topics have varied now and I feel I missed the original intent.

Can you please clarify the idea,  such as what the framework might do (extend and override? call .run() only? call once per frame instead? nothing?) and what the std lib aims to provide?

In my current understanding, the only goal is an API that would allow end users, libraries, or language features to rely on: i.e by calling `MainLoop.add` they will be know that it will be updated by _something_ for them, and will be guaranteed such by the MainLoop construct so they can progress over time.

**Comment by back2dos** [comment]
@RobDangerous Then I say this is an ivory tower debate around some abstract whatsoever that does not even have a clearly defined goal and I highly doubt that anything good will come out of it. Which leads me back to my original stance: put it in a haxelib. 

If this is really about what @underscorediscovery said, then this solves it:

```haxe
package haxe;

interface Scheduled {
  function cancel():Bool;
}
interface Scheduler {
  function add(task:Void->Void):Scheduled;
}
```

Every framework can provide implementations at will. Ones that are locked down with FPS or ones that are tied into the actual native loop or whatever. Slamming this into some static class that frameworks need to shadow will simply cause conflicts and thus incompatibilities. It also takes away control from the user.

Code that can leverage a scheduler can be written against an interface such as the above. Users can still decide to then run it on an eager scheduler if they want the result synchronously or what not. Also, if this is meant to be minimal, then prioritization is out of place. That should be built on top because it's not something you want to have replicated in every single implementation.

**Comment by hughsando** [comment]
I can think of only a couple of reasons for some kind of standard main loop.  In console apps when the developer starts a timer or thread an expects the program to hand around while these are pending.  It is not clear weather the console loop should wait for threads to finish - I think it will depend on application.  To this I might add async io, if you want a generic async haxe.http.
So I propose  entering a blocking loop at the end of main, unless a "asyncWake" callback has been registered (see later),  while there are timers pending.  To be cross platform, the timer callback should get run on the main thread.  For browser + flash the loop is only notional, like it is now.  

You could also consider adding a sys.Thread class which can take an extra parameter to the create function, "waitForTermination", or maybe "Scheduler.createThread(callback)",  with this set, the sys.Thread class wraps the user function with another function that increases/decreases a "current waiting thread count" and pokes the even loop when the count drops to zero and otherwise prevents the main loop from exiting.

Any kind of "onFrame" callback, or even a "tick" presupposes a concept of frames or at least keeping the cpu awake, and should not be part of the standard.

For async io, eg http onload/onerror, the main loop would also block on console apps while there are pending requests.    The easiest way might be to simple spin up a thread-per-request, and then use the thread-exit mechanism described above.  libCurl allows you to combine all sockets into a single thread - haxe http could do this too, or use a poll.  For browser, this is handled natively already.  For flash, you would create background loaders - no need for main loop.  For node, I guess they have a pretty good solution here.  For gui frameworks, I'm not too sure.  Nme solution here is like the timer solution, that is "poll" and "get next poll time".  With no timers and no callback pending,the poll time is infinite.  Other events (such as mouse clicks) cause the next poll time to be recalculated, and other events (eg, threads) can generate asyncWakes.  async io and timer callback happen in the "poll" callback to ensure the thread is known and consistent across targets.  How the async io handlers register themselves I do not have much of an opinion about, but nme uses a "poll Client"
https://github.com/haxenme/nme/blob/master/nme/app/IPollClient.hx
But it should be pretty simple to adapt.  This will need to be platform specific, because this does not make any sense for browser or flash.  I guess the haxe.http class will look completely different in these cases.

So I propose a  solution, say,  haxe.Scheduler, with "timeToNextPoll" and "Poll", and "asyncWakeCallback" for the frameworks.  Adding a timer, async io or a thread finishing cause the nextPollTime to change, and an generates a asyncWake callback.  For targets with built-in loops, these functions simply do not exist.  For console apps, the main loop is blocked while there are pending events.  For frameworks that want to control the main loop, they must respect the "timeToNextPoll", if it exists and call "poll". They also need to register a "asyncWake" callback, that allows newly scheduled timers and async io to get serviced when there are no other callback pending.  Developers wishing to schedule a poll or keep-alive for their own async io will need to be aware of the target, but can implement the pollClient on framework/console targets.

This should give timers and async io on all platforms and thread-waiting for console apps.  I think adding thread-waiting to frameworks is not great, but could simply be achieved via a Scheduler function "areThreadsRunning()", and cancelling the window-close if you really want.

**Comment by ncannasse** [comment]
@back2dos I think you need to read a bit more before writing :) The example code you are showing is exactly supposed to enter an infinite loop, at least on a "system" platform such as neko or php. You're registered a new event loop and it will be run until stopped. Of course on JS/NME/Lime/Snow this would not freeze but instead refresh the window then call the event depending on the screen refresh rate. Read more about why it's useful in Hugh comment.

@hughsando yes I think we might want to have poll'able events as well, but can't this be built on top of what I'm proposing? For instance if you use epoll() for managing your sockets, you would epoll() every tick(). You can eventually epoll() infinitely if there's only one MainEvent registered (which would be the case if you're only waiting for a Http call without any display occurring). Or you could have threads that do that in the background, then add a MainEvent in order to make sure the callback is run into the application thread.

Please clarify your haxe.Scheduler API proposal as a GIST, we should not assume that we can have multiple threads running on the platform.

@underscorediscovery I would like this to become a standard that Haxe users can rely on, so for example one that uses Snowkit would be able to use this API and get things called as specified (the current discussion is about discussing the specification), so yes in order to work correctly Snowkit will have to support it since using the default implementation would freeze the application as Juraj cleverly noticed.

**Comment by ncannasse** [comment]
BTW, I *really* want that we collaborate at finding a solution that is acceptable (if not perfect) for everybody, so if you have issues with the API proposal I made, please think about it and propose your own, or else you will be bothered in a few months when users start using it and ask you to add support for it in your framework :)

**Comment by back2dos** [comment]
@ncannasse Thanks, I think I've read enough. I see my feedback is not welcome. Have fun.

**Comment by ncannasse** [comment]
@back2dos don't take it badly, your feedback is welcome as long as you're trying to understand what we're trying to do, and is being constructive.

**Comment by hughsando** [comment]
Juraj, I think you have been distracted by "MainLoop" which should be
considered only one possible solution.
Instead, perhaps you could take a step back offer an opinion on the problem:
How can the haxe std libs, particularly http, socket io, database io,
File.getContent, directory.read, as well as custom libraries, such as "on
filesystem changed" interact with a command line main loop, a browser main
loop or a framework main loop (nme/opengl/kha/heaps/waxe/...) in a
consistent way.

I'm thinking browser (js/flash) will be significantly different to the
other targets, but I could be wrong.

**Comment by ncannasse** [comment]
@hughsando the only difference I see with browser (and maybe mobile as well?) is that we need to actually return from our haxe main() after registering ourselves into the plaform loop that we don't own, whereas on other platforms we need to keep running until we're "finished" (which my MainLoop defines as all events have been stopped, but that could be all windows have been closed, all threads have terminated, etc.)

**Comment by hughsando** [comment]
Nme registers an "onWindowCreated" callback and exits - openfl used to, not
sure anymore.  Waxe exits too.  Flash/js obviously so.
Command-line is the only "framework" which needs to block after the main
AFAIK.

Either way, enforcing that a framework sit in THE loop, makes is completely
unusable for me (nme,waxe).

**Comment by back2dos** [comment]
@ncannasse I am sorry for not having understood your code. It's just that virtually all major runloop implementations out in the field will execute scheduled callbacks once. This is true for node's `process.nextTick`, for the DOM's `window.requestAnimationFrame`, for Cocoa's `NSObject:performSelectorOnMainThread`, Swings `SwingUtils.invokeLater` and what not. I will leave it open to speculation who's lack of reading is the cause of my inability to understand your code. Let me leave it at reemphasizing that with this particular choice, it will be close to impossible to integrate decently with most environments, which is a point that others have raised. Please do take that seriously.

That is not my point however. My point is: this - whatever "this" is - is a horrible idea. This thread is filled with implicit assumptions none of which hold. The result will cause more harm than good. I do not have the time or energy to discuss that on some theoretical level. Problems will occur naturally once you try to integrate this supposedly abstract API with other systems. My proposal was thus to put it in a haxelib to contain damage (and also - instead of adding to @Simn's frustration over pending issues - have a separate tracker to deal with all the problems that I predict you will have with this) and elevate it to a standard once it is actually shown to be usable. I have also pointed out an alternative to baking this into the compiler. Both proposals are not only constructive, but even directly actionable. And I would argue that they are pragmatic too, leading to tangible results more quickly.

@hughsando As for your concrete question, I think it is a very interesting one, but it is remotely related at best. Tautological as it may seem, the short answer is: for asynchronous IO you need abstractions for asynchronous IO - and *not* abstractions for run loops. *Interfaces* that act as the asynchronous counterparts to `haxe.io.Input` and `haxe.io.Output`. Some implementation may indeed be backed by a synchronous input/output and a way to offload IO to a background thread and have the result be dispatched back onto a main loop of sorts. Other implementations may simply rely on what the native runtime environment provides. I [have published work in this area](https://groups.google.com/forum/#!msg/haxelang/2gRl_zCc2-0/VHUUwTp6AgAJ) a few months ago that provides a homogeneous TCP and HTTP API for nodejs, java and neko. We can gladly discuss what little I was able to learn and the countless things still left to do, but I don't think this particular issue is the right platform for that.

**Comment by hughsando** [comment]
I have put an outline here:
https://gist.github.com/hughsando/91a19a61a82bf8b7a9e6

The basic idea is that frameworks can call "poll" and "getNextWake"
directly and the console can block and do it automatically.

If you do not wish to add the thread stuff, we can just block if there are
timers pending - most frameworks already support the haxe.Timer api,
although this does not allow for a Zero-CPU solution.

As for haxelib vs std lib, incubating in a haxelib seems like a good idea,
but if you convert haxe.Timer over to it, maybe not.

**Comment by back2dos** [comment]
On Sun, Mar 6, 2016 at 1:34 PM, Hugh Sanderson <notifications@github.com>
wrote:

>
> As for haxelib vs std lib, incubating in a haxelib seems like a good idea,
> but if you convert haxe.Timer over to it, maybe not.
>

During that period such a haxe.Timer implementation could simply be part of
said haxelib.

**Comment by Simn** [comment]
I think that

* we (the Haxe Foundation) should manage it so that we end up with a centralized solution that works for everyone, and
* it should be a haxelib so that <strike>I don't feel like I have to maintain it</strike> we can make independent updates.

**Comment by underscorediscovery** [comment]
- I think the haxelib idea is better 
- I agree with Hugh and Juraj and have stated: as described it is not something I feel will solve the right problems. 

It is overly invasive for me to feel comfortable with - at least as described. Ceding control over something as fundamental and crucial to cross platform portability of the system level loop cycle is just not an option for me, nor nme or others which has been stated as well.

**Comment by dazKind** [comment]
IMO, put this into a simple haxelib. if it proves to be a viable thing that people need and use you can easily merge it into std.

**Comment by ncannasse** [comment]
@back2dos saying at the same time this is a horrible idea AND that you don't have enough time to explain why, without giving even the slice idea how why you would think that, is not very nice to say the least, and definitely not constructive. Not that I want to enter a heated debate on this topic, please simply make a contribution by proposing an alternative that can be discussed.

Choosing to go haxelib or haxe std will depend on how complex the code is and how likely it is to be changed in the near future. This will only happen after we have decided on it. I didn't happen this way but if everybody had agreed with my first proposal I don't see what value it would have to put such simple piece of code on haxelib.

We want a solid piece of foundation code that is shared across frameworks and that people can build over.

@underscorediscovery the idea is not to cede the control, but to let users add their own systems to it, with a minimal specification that the framework will be requested to follow. The framework will of course have its own implementation that is optimized for its own subsystem. I don't see how that can be something bad or that doesn't solve problems.

**Comment by ncannasse** [comment]
@hughsando thanks for submitting your idea. I find it a bit too much oriented towards threaded/native systems, I'll try to improve over it.

Regarding your clients management: using a double linked list as I did in my MainLoop ensure that removing clients that are currently / will be run is correctly supported.

**Comment by ncannasse** [comment]
@underscorediscovery could you also propose something that would work for you?

**Comment by hughsando** [comment]
I don't imagine adding/removing will happen very often.

But as far as threads go, you can assume these are just "#if" ed out and
you can ignore them on non-native targets.  BUt the whole thing practically
disappears on browser targets, so what problem are you trying to solve?

 I do think you have to run the callbacks on the main thread or else you
will complicate the developers code.

**Comment by ncannasse** [comment]
Here's a second version that allows to delay() an event, and which allow another thread to wakeup() the main loop.

https://gist.github.com/ncannasse/8048e759daa6cea3ce8c

There's still a good number of things to do wrt threads safety, such as protecting the events list, we can add these afterwards, nothing hard there. The code is just to give you an idea.

@hughsando yes, you want of course have the callbacks run in the main thread. For this in my MainLoop you simply add an event with no delay that will be run once and stopped.

@back2dos the reason why events are run in loop by default instead of only once is to reduce allocations. For one-time event you stop() when its called first. For many-times events (setInterval like) you have nothing to do.


As an integration for instance in Snow/Luxe @underscorediscovery, that would mean that you have to include haxe/MainLoop.hx in your framework, redefine run() only, optimize a few things that might be relevant, and somehow make sure that tick() is being called

The problem I'm trying to solve is to have a crossplatform way to develop async code. ATM we lack the constructs and it requires #if'ing for each framework available, which is not a good. I want this to be:

a) lightweight
b) efficient
c) adopted by all frameworks

**Comment by delahee** [comment]
I think Hugh has got a nice lead here. If we could simplify a bit, it would be nice. Nicolas approach is interesing too but I kind of worry it is too straightforward in this Activity/Delegate/requestAnimationTimer hostile world.

As to haxelib or not, despite what I heart, my brain says the std.mainloop could go in an official haxelib just like spod &  dispatch should.

Annnnnnd since there are mostly frameworks writers here ( which are sparse users of third party libs ) and since tivo was very vocal about main loop, let's bug them :)

ping @kulick

**Comment by Simn** [comment]
You're probably looking for @bjitivo.

**Comment by underscorediscovery** [comment]
I have earlier,

> If the user would call MainLoop.tick() and not run() once, this would work already because of platform differences.

The second one with the tick fits closer to what makes sense in practice. Who is supposed to call run? Is it called implicitly at the end of main()? If so, I will override run with a blank function and retain the control needed. Calling tick explicitly to update the events is perfectly fine from my view - it is the least invasive and retains user control over it being there in the first place.

To me it makes sense because:
- CLI/other use case can add explicitly an event into the queue 
- Std lib can add events into the queue via API calls like sockets etc
- Frameworks can progress the queue themselves from whenever/wherever it makes sense to do so

It's important then that tick be the only point of contact and handling, i.e there are currently structures inside run that the user would have to replicate in their own code and would be critical to it functioning like the mutex. I would move these into the tick function so there is only one point of contact that should be "called often" and no other strange coupling.

You could condense the delay as a parameter to `add(delay, fn, priority)` to simplify usage.

**Comment by waneck** [comment]
I really like Hugh's suggestion. It has the concept of main loop sleeping / awakening, and it allows timed events in a way that doesn't involve infinite looping until the event it ready :)

Coincidentally, I've been working on a scheduler for the native targets, so this is very fresh in my mind. I think that Hugh's solution covers most of it, which has to do with being able to awake it when a specific event has happened (e.g. epoll_wait has returned), being able to run events on the main thread, and allowing timed events to happen.

There are some more complex cases where we might have multiple threads running epoll for example on the same set of events/sockets. This allows very efficient use of one selector per cpu, but I can imagine it might be outside this scope.

There are also some possibilities of having multiple loops inside the same application.In my own use case, I was thinking about allowing an asynchronous class (e.g. `Http`) to receive an optional `Scheduler` argument. Otherwise, it uses the default.

I'm a little worried in this sense about how each implementation will setup its Scheduler. It this based on classpath? This might get tricky depending on the scenario and set of libraries you're using.

**Comment by ncannasse** [comment]
> The second one with the tick fits closer to what makes sense in practice. Who is supposed to call run? Is it called implicitly at the end of main()? If so, I will override run with a blank function and retain the control needed. Calling tick explicitly to update the events is perfectly fine from my view - it is the least invasive and retains user control over it being there in the first place.

Yes that's exactly it. Actually I think we should not have run() there but in a separate class `EntryPoint.hx` that would simply do `MainLoop.run()` so you only have to redefine `EntryPoint` without having to sync the `MainLoop` code in case we make updates.

@waneck regarding multiple event loops, we could make MainLoop an instance + a static singleton, with current static methods that would use the singleton by default.

**Comment by ncannasse** [comment]
Regarding Hugh implementation, I have a few questions/comments so we can combine with my updated MainLoop. I've tried to list all the differences:

 * I tend to prefer using double linked list for clients implementation, which would allow for partial locking when manipulating the list instead of fully locking it. This should allow more thread-safe iteration on the list, and cause less copies/allocations when pushing/removing from the Array. Any disagreement?

 * Is getNextWake necessary or a fixed delay such as my MainEvent.nextRun good enough? I can't imagine right now an event that would change its delay dynamically. nextRun is more efficient but requires a class, not an interface (see below)

 * any poll() vs tick()/call() vocabulary preferences ?

 * using interface IScheduleClient VS closure (te latter feels more "Haxe" to me) 

 * after giving it some thoughts, threads API seems good to me. However I think we should keep it private, so the platform thread implementation is responsible for calling it.

* should our default haxe.Timer implemenation use a single Timer Manager as Hugh suggests or should all timers be registered as separate Events as I thought would be possible in my MainLoop v2 ?

**Comment by Justinfront** [comment]
I am curious about the javascript loop in Nicolas's gist
  https://gist.github.com/ncannasse/8048e759daa6cea3ce8c
For browser js, is a css3 injected loop better in some situations?
  https://github.com/Justinfront/divtastic3/blob/master/src/core/CSSenterFrame.hx
and if so, what would be the setup for me to override the conventional inbuilt approach, is this going to limit developers present and future and even past by impacting old code?  For instance on as3 you assume onEnterFrame but again maybe it's not quite the same for Red Tamarin, maybe it has a better setInterval or something.  On java swing if I wanted a couple of threads to handle different aspects of the app is that setup, say one for render and one for interaction events? 
Lastly what happens when we are mixing Haxe code with platform code will this setup get in the way in how we are forced through internal politics, or historic reasons from using Haxe smoothly, for instance loading as3 into haxe into as3 movies with different types of domains. Maybe it's all cool in all situations.

**Comment by ncannasse** [comment]
@Justinfront the idea is to be able to simply redefine the run() method, so one given framework can hold the loop and do its own stuff, still calling tick() so that crossplatform events are correctly handled. I really don't think this will cause issues with AS3 domains or other scenarios.

**Comment by ncannasse** [comment]
I have put a new GIST that separate between MainLoop (cross platform) and EntryPoint (framework-specific). I think that Hugh support for threads would go in EntryPoint.
https://gist.github.com/ncannasse/3614dd6b6e7f7242b60d

**Comment by underscorediscovery** [comment]
- I think your notes on `run` make sense. Maybe it could be like how `trace` is reassigned, since run is a single call and not every frame it's fine if the call isn't the most direct/efficient.
- poll/tick either vocabulary is fine to me, I also like `update` or `process` but poll is common in event loops
- The only current thought I have on the next wake model is that it doesn't try and do any frame timing and just runs as fast as it's called (not related to delayed events timing obv). The only (very much) older implementation of the nextwake that I encountered in nme wasn't precise/correct and would create unstable timings. Timing the calls should be in control of the user when it's tick is run.

**Comment by Justinfront** [comment]
Nicolas maybe you addressed my main point, or missed it?  So I will clarify:
How can I swap
```code
		var window : Dynamic = js.Browser.window;
		var rqf : Dynamic = window.requestAnimationFrame ||
			window.webkitRequestAnimationFrame ||
			window.mozRequestAnimationFrame;
		rqf(run);
```
For this code
```code 
var s = Browser.document.createStyleElement();
        s.innerHTML = "@keyframes spin {  from { transform:rotate( 0deg ); } to { transform:rotate( 360deg ); } }";
        Browser.document.getElementsByTagName("head")[0].appendChild( s );
        (cast s).animation = "spin 1s linear infinite";
        loop( 60 );
```
If I found that CSS3 loops were running faster, or even decide based on browser or device?

**Comment by Justinfront** [comment]
mmm I am trying to work out if they are doing the same or if there are differences in how they would work and in resources used, mine is 60fps, but I don't know what frequency the first one runs at would it be called lots more and result in more resource because tick is going to get called lots even if only decides internally to do stuff every so often, plus you have all that extra Timer stamp code run each time?  I am not sure if my point is good or bad but I suspect it highlights the inflexibility?

**Comment by Justinfront** [comment]
So I would call tick myself, but then the saving of the CSS3 doing the 60fps calculations in hardware is wasted?

**Comment by nadako** [comment]
Minor impl note: from HaxeQuake I remember having problems with RAF when tab is inactive or something, so I sticked to `setInterval` for that. I guess RAF is designed for rendering, not for general event loop systems.

**Comment by hughsando** [comment]
This seems reasonable - you still need some way of waking the framework up from a deep sleep.
Also. I think the js reference implementation should not request an animation frame until there is an event to run - we must be very conscious of allowing a zero-cpu solution.  You may also like to use a js timer in the case where the next wake is more than say 1 second away - since the library code only needs to get written once, may as well make it as efficient as possible.
Lists vs array is an implementation detail, and I do not care either way as long as it works well.  If you have a struct rather than an interface (as you do) it means you can change implementations without anyone caring, which is good.  Similarly, combining timers vs not is just a detail?  If haxe.Timer extends MainLoopEvent,  you may be able to keep the same Api (new Timer().run).  Although, this may break frameworks that already override Timer and do this kind of stuff themselves.
I guess you would also handle threads by pushing an event, and then poking a wake event?
So you would still have "MainLoop.runOnMainThread(Void->Void)", it would just create a struct and push it on the list.  Some kind of "runOnce" to auto-remove it from the list when run.

Sven, the next wake time is just a suggestion - how the framework combines this with its frame-rate (eg, only poll the main loop a 60hz) is up to it - as is the case with js request animation frame.

I do not like the name "tick", mainly because it makes me feel like you should call it a 60Hz, but I could live with it.

@waneck, I would like to see http (and sockets in general) implemented with a single thread blocking in a multiple select call on platforms that support it.  My

**Comment by bjitivo** [comment]
We did this already: https://github.com/TiVo/activity  ...

... and presented at WWX: https://github.com/kulick/wwx2015/blob/master/WWX2015_ActivityHaxelib.pdf

Feel free to use or not use as much of that as you want.  I would love for the program's "main" to be the first "activity" started up by the generated code automatically, with no need to call Activity.run().  Then programs that never care about using concurrent programming can look exactly like they do now, but anyone can create more activities from main if they want too.

**Comment by hughsando** [comment]
I see the activity library as a client of the MainLoop, not the other way around. This allows win32 apps to sit in "GetMessage/MsgWaitForMultipleObjects" and android apps to wait for java events from the view.
The key api being something like:
```
  // For frameworks
  static function setAsyncWakeCallback(Void->Void);
  static function runEventsAndGetSleepTime():Float;

  // for clients
  static function asyncWake();
  static function addEvent(event:MainLoopEvent); // runAndGetSleepTime + remove/cancel
  static function changeLockCount(inDelta:Int); // Maybe add a dummy event instead
```
So, as a client of the MainLoop, the activity library would add a single event to the main loop and do all its soon/later/running inside a runAndGetSleepTime, return how long it wants before being called again.  If it spins up a thread to sit an a message queue, it can call asyncWake when it gets something.

We then have one of several frameworks - command line/ browser/nme/openfl/waxe which use the first two functions where appropriate.
Doing it this way extends the lovely activity async socket code to all frameworks that want to implement the 2 simple api functions.  So I guess the question is, can you work with just runAndGetSleepTime and asyncWake?

**Comment by ncannasse** [comment]
@Justinfront I think requestAnimationFrame is the standard for JS, which don't really need more than 60FPS refresh rate anyway. We can still swap the implementation but I think that's the best for now.

**Comment by ncannasse** [comment]
I'll try to update my MainLoop proposal with code regarding thread handling and concurrent access, and have a working Neko implementation. Is there something else to be done?

**Comment by ncannasse** [comment]
Putting that forward to 3.3.0-RC1, still waiting for @jgranick input on the topic

**Comment by mcheshkov** [comment]
I'd like to add some words, mostly influenced by Qt and libuv design. I'll be verbose, and start from beginning, since I feel lack of consensus.

There's an event loop, as a whole, and there can be many of them. Each loop have some methods, most important is `run` that blocks until there's no more _possible_ events pending, or until stopped, `stop` to make loop exit at next tick, and `post` to add event. By possible I mean, that there's no thing to wait events from, even if there's no pending events now. 

Each thread can run different event loop, and nested loop are prohibited (maybe it should not be, but I see no point in blocking call to nested loop `run`, instead of posting same events on your loop). Communication with loop from other threads should be message queues. If one have rendering and network IO, one may want to split that in separate threads, posting events to each other.

There's a default event loop, for flash / js in browser / node.js / etc. it's native loop, for other platforms it can be static initialized in library.

There's event sources and event listeners. Event source is not an object in OOP sense, it's more implicit thing, like "event loop is idle", "new frame is requested" or "socket have new data to read". While code that push events to loop can be external, some of that sources should be considered in implementation. Binding listeners to event is completely external to loop, it cares only about receiving event with callback and calling it.

There are different kinds of events, that may be handled differently by loop.
Idle events should run on next loop tick, keeping CPU hot, preventing loop from breaking but still allowing to do stuff async.
Timer events should run after specified time passed, so other activity may interfere with that.
IO, draw or any other underlying system facility - depend on that system much. I.e. `epoll_wait` may accept timeout, and block for limited amount of time.

If you have nothing to do, but IO, you could wait some time. If you have pending timer, than you should not wait more than that. If there's idle callbacks, or other stuff, you should only check for pending events, setting timeout to 0. If there's only timers, no need to waste CPU, let it (nano)sleep. PostEvent should, of course, wake loop.

All of that is `run` implementation details, that seem important to me. Some of this details may be pushed away, i.e. to different thread, like sockets polling thread, that blocks on epoll and post each socket event back to loop thread, or timer thread, that just sleeps right amount of time, wakes on new timer, and posts events back. In that case loop implementations may be pretty short, wait(could be busy wait on spinlock, could be condition variable) on concurrent queue, run callback, loop. I don't know how that will affect performance.
Similar idea can be used with flash/js, but instead of thread one will listen on some native events, and post them to implemented loop, and from there they will be called. So implemented loop will act as a hub, but maybe that's useless overhead.
API would be like `TimerEventSource.delay(eventLoop, callback, delayAmount);` will push new timer to timer thread, and later from that thread new event will be posted to said loop. Or `socket.readable(eventLoop, callback);` will bind callback to state in io thread, and on state change thread will post event. `eventLoop` argument could be optional, default loop will be used if none passed. It allows both custom object tagging in event source implementations, and common event loop api.

With all that in mind, I think that Haxe could provide (in stdlib or not) API for event loop and test suite to check implementation behavior. Maybe default almost-noop implementations for platforms that have own loop, and even default proper implementation for ones that does not. API should cover immediate events - there's no need to wait, i.e. for animation frame just to make async API async, like in `then` on already resloved Promise. API should cover timer events, so possible implementations could not waste CPU in _some_ cases. API may cover other events, and that's too abstract, but I don't see a clear way to put RAF, XHR, WM_PAINT, epoll, IOCP and all other event sources in single cross-platform API.
That's much like @back2dos `Scheduler` proposal, but covers more different types of callbacks, that may influence efficient loop implementation.

Regarding examples:
I see @ncannasse one more concerned on implementation, while what library-framework developer want to make library-loop separation is interface only. @hughsando is more API related, but `getNextWake` is clearly tied to implementation, since there's no such thing in browser. Also `Scheduler` is all static, and so only one event loop is allowed per process, but that's easy to fix.

**Comment by jgranick** [comment]
Lime does not use Haxe callbacks in its main loop, in order to get smoother frame times all of the timing is handled on the native side.

I am struggling to understand the use-case?

Of the platforms that Lime supports, Emscripten, Flash and HTML5 are the most "exotic" when it comes to a main loop, but on none of them does Lime initiate a loop immediately.

If there is a favorable reason for using the `activity` haxelib, for example, I would be happy to support an `#if activity` conditional in the setup code to work more favorably, similar to what we do for Munit.

Is the desire to provide a main loop implementation for Haxe projects that do not use a framework? Is it to add code before a main loop is started? Is it to get update/tick events during the loop? Is it to intercept events? Perhaps if I understand the use-case better

**Comment by hughsando** [comment]
It is important not to choose "the best" way of doing an generic event loop
(post-to-multi-queue, activity-lib) but provide a means by when ANY event
loop style can interact with ANY framework.  So there are 2 parts - 1. what
does a framework need to do support generic async-ish libraries, and 2.
what does an async library need to do to be supported by a framework.  To
be explicit, i would say:
frameworks:
  nme, lime, waxe, qt, sdl-based (snow, custom), flash, js, native win32,
command-line*
async-ish libraries (MainLoopClient):
  timer, http, socket io, audio player, libuv, activity, file system
watcher, MainLoopThread

Now I think it is important to split command-line into the framework class,
and let the main loop library provide an *optional*  default implementation
here.  But this is different - the other frameworks will not want to use
any code in this class.  I think is is where some confusion with the
"MainLoop" comes from.  There is "LibMainLoop", which is what I want to
talk about, and is for "integrating with the main loop" and
"CommandLineLoop" which is just one implementation which will not be used
at all by nme.

Anything that does not involve the main thread can be handled differently -
does not need and framework support, so we should ignore it for now.
Although, I will add that some thread activity may keep the command-line
framework alive, but adding some kind of dummy client seemed like a good
idea.

LibMainLoop should not block at any time - you will mess with the
frameworks timing.

So what does a framework need to do to support an asyc -library?
  1. callback on the main thread at later time (eg, timer, polling)
  2. callback on the main thread due to async event (eg, filesystem change
event)
  3. provide some kind of activity state (active, inactive, terminating)

What does a MainLoopClient need to do?
  1. Register/unregister
  2. Request async callbacks
  3. Say how long it has until it wants a timed callback
  4. respond to activity states.

The command-line framework would block until there are no more clients, and
the implementation details can be separate.

I would see something like 7 public functions corresponding to these 7
tasks in the API.

So, to answer Joshua's question its a way for any MainLoopClient to
interact with any framework.
Lets take haxe.http as an example.  This class would use a MainLoopClient
to requests a framework callback to dispatch its data/error info.  This
will then work commandline, nme, waxe, lime, ...
Same goes for a hypothetical file-system-watcher class.  This does not need
to be tied to lime, it could live inside the commandline framework as part
of, say, a webserver that also receives http requests.

I think you will agree this is useful.
And again to re-iterate, this LibMainLoop should not enter any
"while(true){ }" loop, except in the command-line framework class,
otherwise it is unacceptable for nme and waxe and I will not use it.

I do not think the 3 framework functions i have proposed are too hard.  1.
is just a generalization of the timer class. 2. (async callback) is doable
and desirable, if not already done. 3 could probably be ignored if you want.
The client functions are probably more for client writers to discuss, since
frameworks do not need to know about them.  But some kind of set of
MainLoopClient classes would seem a logical implementation.

Hugh

**Comment by jgranick** [comment]
Thinking of Haxe core classes, it would be nice to be able to be able to somehow provide Haxe with the current milliseconds and to have a callback to enable that code to work.

Other than the milliseconds, this seems similar to the `requestAnimationFrame` API, it assumes no knowledge of the underlying implementation, but allows that "heartbeat". 

Is this heartbeat enough for most uses?

**Comment by hughsando** [comment]
I think "heartbeat" necessarily involves continuous use of cpu, and we definitely need the api to allows for a "zero cpu" solution.  That is not to say that a framework may not instead choose to poll on a heartbeat - eg, the js implementation may simply poll on requestAnimationFrame, and that is ok.  In fact, since there is no owner of the js loop, a js framework may not need to actually do anything.  eg, an async https request will still work on lime+js even is lime does not know anything about it. This is like how the haxe.Timer class just works on flash.
But waxe will want to sit in a "GetMessage" loop on windows if nothing is pending, and not have to set a 10ms timer that does nothing.
So a lime heartbeat will be all that is required to support this on some level.  If you want a more efficient version, you may choose to implement "asyncWake" at a later date.  The MainLoopLib can support both modes easily enough - just poll the library irrespective of the requested interval or any async wakes.

I see this a a standard for how async libraries can interact with frameworks.  If the haxe classes (eg, http, timer) are moved over to only use this I would gladly ditch the nme timer implementation.

As for sys.io.File.getContents on android, I think it is easy enough to override this with your own class path.  The haxe.io.Input classes are well-defined enough a user of your library need not ever know.
I think this is a different story, but maybe what you are talking about is something like:
haxe.Framework.getContents, which by default calls the sys.io version, but via some registration mechanism,  allows a framework to present some kind of virtual file system.  I like this idea as a separate initiative.

**Comment by jgranick** [comment]
So "keep me posted" and "call me, maybe?" callbacks. If this were formal, I would prefer if there was a core implementation class such as `sys._framework.SysTimer` or something that lives behind the user class, hidden to the user. If it were meant to be overridden, we could implement these functions without overriding the whole classes as we do now

**Comment by ncannasse** [comment]
Ok, I've setup a first implementation proposal, please review it there:
https://github.com/HaxeFoundation/haxe/pull/5017

 - MainLoop is the "user API", the one that should not be modified by the framework
 - EntryPoint is the platform default implementation, it can be redefined by any framework that is willing to integrate this API. I have provided implementations for JS, CPP, Neko and Flash.

Regarding the MainLoop API:

 - addThread will create a new "thread", if the platform does not support threads it will run the function at the next event loop.
 - addEvent will add a new event. The event will run until it is stop(). It can specify its nextTick if he wants to get delayed, or can leave it to null to be called as regularly as the framework/platform allows.
 - runInMainThread() will push a function to be run into the main thread. Please note that addEvent is not thread safe, so a framework which want to register events should wrap the addEvent into a runInMainThread. Same of MainEvent.stop()

If MainEvent/EntryPoint are referenced, an EntryPoint.run() call will be added immediately after the Main.main() by the Haxe compiler. We could improve this by checking if MainEvent survives DCE.

Looking forward for your feedback/patches, I'll merge for RC.

**Comment by Simn** [comment]
I still don't think this has to be in the standard library... especially not while it's still being discussed.

**Comment by ncannasse** [comment]
@Simn it is absolutely necessary to have this in standard library since it requires both compiler support and standardized API. We can still provide updates to the API on haxelib between 3.3 and 3.4 if we really need to introduce additional features in the meanwhile, but its place is in haxe std.

**Comment by Simn** [comment]
These std-specific hacks in the compiler sources are terrible. I want to get rid of them and you add more instead... This should be designed properly and come with a suitable interface, not be hardcoded to a specific type.

Standardization can be achieved with a haxelib as well. Whether or not something is in the standard library or a HF-maintained haxelib is just a matter of distribution.

**Comment by ncannasse** [comment]
@Simn we can always improve the implementation details afterwards, I think we need to advance first and get things done. A lot of people (including me) prefer to keep the number of haxelib dependencies to the minimum, having things in haxe std helps there.

**Comment by Simn** [comment]
I think a lot more people prefer to be able to receive updates and bugfixes easily without having to wait a year for a new Haxe release. It's one thing to add a proven concept/implementation to the standard library, but what we have here is still very unstable and experimental. The natural approach is to work on this as a haxelib, then maybe add it to the standard library once it stabilized.

**Comment by Justinfront** [comment]
I tend to agree with Simn's caution on this, the wrong javascript implementation could be quite harmful to Haxe, I suspect for C++ it's less likely to be quite so important to get perfect. My understanding is this change would effect pretty much every codebase it's not something that a user can easily not use under the current proposal?  You invented haxelib's for these types of decisions !

**Comment by ncannasse** [comment]
@Justinfront if you don't use MainLoop, it will not affect your JS code.

**Comment by jgranick** [comment]
It is interesting to see better what you have in mind. Would it be possible to detect if a user has opted into using a MainLoop or not?

Similar to the `EntryPoint.run` that you have set up (but after creating a Lime `Application` with the user's configuration), the application starts and ends with `var exitCode = application.exec ();`

However (and this is the key difference) all events from then on are signals, such as:

    application.onUpdate.add (myUpdateCallback)
    application.window.onResize.add (myResizeCallback);

There is no loop to intercept. I would prefer the native loop to remain in native code, at least for default users. We could make concessions to push more control into Haxe, but I would prefer it be opt-in rather than required by default.

This is why I was wondering about use cases, I was not sure when a user would need to control something lower-level than this

**Comment by ncannasse** [comment]
> Would it be possible to detect if a user has opted into using a MainLoop or not?

Yes, if we detect that MainLoop is not DCE, then it means it's used by the application

>  I would prefer the native loop to remain in native code, at least for default users.

It is fine, you can simply add your definition of EventLoop in OpenFL, with an empty run(), and have your OpenFL native loop call EventLoop.processEvents() (or something else) as you wish.

**Comment by hughsando** [comment]
I think this is very good.  I would say that "runinmainthread" should also
call wakeup at the end.
Almost all of the "EntryPoint" code I would reuse in NME and waxe.  The
only function I would want to replace is "wakeup", so an architecture that
did not only use class-path-override would be good.  Or make the api simply
use an instance so I can override this one function, or use dynamic
function for wakeup or whatever.
Class-path-override is nice because it will work with old versions (class
will never be missing), but some split here with a "default implementation"
would be good for code-reuse.  You get the idea.

I am now thinking an std lib would be good.  This prevents multiple
standards - a very serious issue that should not be underestimated.  And I
am quite happy for hxcpp command-line and nme and waxe to use this.
Especially if then we have and haxe.http.async and similar socket apis.  A
20-line webserver written in haxe would be a compelling use-case.

As for compiler hacks, read/write access to the type of the "main" class
should allow an "onGenerate" macro to solve this issue nicely and be useful
in a broader context (eg, a dummy main that adds an instance to the stage
in nme's flash-like api)

**Comment by ncannasse** [comment]
@hughsando thanks Hugh, I have added the missing wakeup() and also only wakeup() when we terminate the last thread

It's a bit hard to use an instance, because it needs to be created before EntryPoint.run is called. For example we could move all the implementation from EntryPoint to MainLoop (expect run() of course), then simply have:

```haxe
class EntryPoint {
  public static var loop : MainLoop = new MainLoop();
  public static function run() { .... loop.processEvents() .... }
}
```

So your custom implementation would replace it with:

```haxe
class MyCustomLoop extends MainLoop {
   // overrides
}
class EntryPoint {
  public static var loop : MainLoop = new MyCustomLoop();
  public static function run() { .... loop.processEvents() .... }
}
```

What do you think about it?

Also, would like to hear comments of @underscorediscovery @waneck @RobDangerous @jgranick on the proposed implementation. We don't have much time before 3.3 RC, and we need to fix eventual problems before 3.3 final (due next month)

Personally I'm quite satisfied with it since it seems to cover all the expressed needs of this thead.

**Comment by RobDangerous** [comment]
Fine with me. Now please add some insane feature so we can run code in worker threads in js and in regular threads in cpp in a portable way.

**Comment by ncannasse** [comment]
I have made another update: nextRun is no longer Null, and I sort the events by priority+nextRun before processing them, which guarantee that if you start a Timer of 100ms and another of 200ms, then block 300ms doing some calculus, the 100ms timer will run before the 200ms one.

@RobDangerous sure, that could be interesting, I actually have something like that for Flash workers, but it's a bit hard to pass functions pointers, maybe with a bit of macros help, but then you loose your context (this, captured locals, etc.) : it's still an open problem to do it gracefully, because you cannot share values. I guess a WorkerJob class (only serializing an initial state) and adding inter-worker communications is the best, it might indeed be built on top of MainLoop for notifications.

**Comment by ncannasse** [comment]
Ok, last call for comments before we merge. Speak now if you have any concern that might prevent you integrating MainLoop in your framework :)

**Comment by Simn** [comment]
My concern is that your PR currently [fails on several targets](https://travis-ci.org/HaxeFoundation/haxe/builds/120410348).

**Comment by Simn** [comment]
Hold on, this is not a current build.

**Comment by hughsando** [comment]
Yes looks good.  Rethinking how NME might implement this, I can ignore the
thread counting so there is probably not that much code to share.
The only thing - what did you  think to the idea of making a generic
replace-main-class macro, rather than the EntryPoint specific change?

**Comment by ousado** [comment]
There is no reason to limit this to one global class object - please take a look at libevent, libev, libuv, boost::asio etc. - all of them permit several independent loops, none of them force the user to use one global "instance" for all event-y stuff. Especially since this is supposed to be a standard, let's get that right, please.

**Comment by waneck** [comment]
The problem I see with this is that there's no good way on HOW people are going to override this. I can imagine that some library could just add their own file that takes precedence over the std implementation, but this starts to get bad if we have more than one library that do that

**Comment by back2dos** [comment]
Uhm, ok. I apologize for my prior tone. I do not wish to discuss the particular issues of this proposal, as there is certainly not enough time and it also seems that you have made up your mind anyway and can't be dissuaded by words of caution, which is what upset me in the first place. I think the way you trying to enforce it is a reckless abuse of your power, that will most certainly not lead to a standard that the community will be willing or able to adopt, but what's most important is that it is not at all necessary.

I have [made a poc](https://github.com/back2dos/entrypoint) that provides the integration you seem to be wanting:

If `entrypoint.EntryPoint.run` is called explicitly, then no auto-wrapping occurs, otherwise it does. All without relying on DCE, hacking stuff into the compiler or whatever. As always, there is room for improvement, but it's certainly not worse than the changes to the compiler that @Simn has criticized.

FWIW, here is the test: https://github.com/back2dos/entrypoint/blob/master/tests/RunTests.hx#L3

Output with `-D manual`: [`RunTests.hx:21: [main,before,manual,after]`](https://travis-ci.org/back2dos/entrypoint/jobs/121444756#L169)
Output without: [`RunTests.hx:21: [before,main,after]`](https://travis-ci.org/back2dos/entrypoint/jobs/121444756#L164)

With this, I once again urge you to let this mature in a separate library. If you need help with that, I will gladly provide it, but your current line of action is downright wrong and I protest. Make of that what you will. That's all I have left to say on the matter in this poorly chosen context.

**Comment by hughsando** [comment]
I think any method is going to have to resolve something.  Ultimately
exactly one bit of code is going to "sleep" until the next event on the
main thread.

The person who overrides is the framework - either the one who creates the
window, or the one who creates the command line (command line probably
needs no work, although not sure how this works for Node.js ).   The
override system can work - this is actually already handled because this is
exactly the case with overriding haxe.Timer.  waxe and nme both do this.
Waxe wins be design when both are present since it "owns" the main loop (it
created the window)

The other libraries mentioned would be the *clients* of the main loop.  If
you could act as either, maybe you check to see if the developer has given
priority to the other one, and the add yourself as a client instead.
(Again, round about how waxe/nme interact, although the nme build tool
arranges everything).

So the libuv or whatever would be a client of the main loop.  Nme would own
the main loop, and both window and async events can happily co-exist.

**Comment by waneck** [comment]
@hughsando but how would this priority system work? Will it depend on the user including libs in a specific order? I can't think of a way that this would not be fragile with the current implementation

**Comment by hughsando** [comment]
I think booting off a macro is preferable to a compiler change.
But I think initiating by simply using a class (EntryPoint)  (perhaps
indirectly - eg, haxe.http) is better than doing it via a "-lib entrypoint".

**Comment by hughsando** [comment]
If you are writing a framework that must play nice with another framework
the authors will need to communicate.
But lets be practical here.  Exactly one thing does this that I am aware of.
And it is still clear - the guy who creates the window owns the event
loop.  So if you are using a framework's tool to open a window, the
framework's tool can ensure that they are in the right position on the
class path.
Do you have some thoughts on frameworks you have written ?  eg, Game engine
integration.  The game creates the windows, so they would own the loop.. If
you wanted NME to do some UI rendering over the top of the games 3d
display, the game engine would need to co-ordinate this and it should be
easy to ensure they owned the loop.  Nme would be running in "client mode"
where events are pumped in, and nme does not block.

**Comment by ousado** [comment]
To follow up on the previous comment: without the option for several event loops, there's no way to avoid using expensive locking mechanisms like mutexes, even if the application itself has no need for them at all. Even when several threads are used, it depends on the application whether there's any need for locks, whether e.g. inter-thread communication with lock/wait-free queues can be used instead, etc.

The currently proposed design makes that impossible simply by forcing everything to use a single global event loop.

Examples are: stream processors that only share immutable state and hence have no need for locking mechanisms; any sort of server application where each connection/session/room maintains its own state independently from others; any sort of application where different sets of events have different scheduling/prioritizing requirements so that one set can simply be processed in FIFO order while others require ordering, etc.

@hughsando's recent comment makes me wonder whether I'm getting something wrong here, is this whole issue about event loops involving GUIs only? Not everything that needs an event loop has a GUI. Rather, GUIs are a fairly special case where the above mentioned scenarios with many basically equal, parallel instances of independent event loops rarely comes up (it seems).

**Comment by ousado** [comment]
@hughsando Also, if this is supposed to be an abstraction over event loops, wouldn't it make sense to allow "clients" (like libuv in your scenario) to expose the same interface this abstraction provides? Otherwise users would have to program against libuv's interface instead, losing the benefit of this abstraction.

**Comment by ncannasse** [comment]
Regarding the various comments:

a) if standardizing things to make sure people can build things in a crossplatform manner whatever the framework they are compiling to is called "abuse of power", then I should maybe abuse more of often of that power so we don't end up with tens of redefinition of something such as haxe.Timer. For instance we standardized Bytes at some point in Haxe history and I don't think anybody complained about it.

b) we don't need several frameworks to have conflicting implementations. As Hugh stated, the window owner is responsible for overriding the EntryPoint definition with its own. Many frameworks already do that for haxe.Timer, so there is nothing new here.

c) having the EntryPoint called by compiler is a temporary measure. Ideally we would like to have some macro code doing that, without requiring the end-user to make an explicit call. EntryPoint.run() will either return immediately on some platforms or will block forever on some other, making it *in all cases* necessary to call it at the end main() anyway. We already have some kind of compiler-specific logic for several other APIs, such haxe.Resource, again nothing new here

d) regarding several events loops, this kind of mechanism can and should be build on top of MainLoop to be crossplatform, or simply use the native platform APIs to do something else before returning from main() if you prefer.

e) regarding mutexes, etc : there's not much of it, and we could surely optimize if you have not created any thread yet. That's not performance critical atm so it's fine. Also, using the MainLoop does not prevent you from using directly the platform specific API

**Comment by Simn** [comment]
> without requiring the end-user to make an explicit call

What would be so bad about that?

**Comment by ncannasse** [comment]
@Simn what is so bad about automating it ? 

Automation doesn't require to put a call to be put at the end of main which is the only actual place it would belong and everybody would have to put it there anyway, leading to many new comers not doing it and complaining that their app exit immediately.

**Comment by Simn** [comment]
I'm not saying that we can't automate it *eventually*, I'd just prefer not to make some compiler hack for this if we can get started without it. That would also remove the necessity to have this in std right away and give us some more time to develop.

**Comment by ncannasse** [comment]
Yes, but it push back the standardization to 3.4, since nobody will be able to rely on it, and i don't see framework owners adding a dependency to a library that is meant to disappear. I prefer that we do something for 3.3, which can be improved later. Again if you have specific remarks regarding the interface, please tell

I don't want much to hear that we should do X better if only we had Y, because that's not being constructive (unless you provide a PR for Y of course :) )

**Comment by Simn** [comment]
> i don't see framework owners adding a dependency to a library that is meant to disappear

What do you base that on? There are at least two framework owners (Juraj and Sven) in this thread who advocated putting it in a haxelib. You seem to be the only one who's adamant about adding it to std right now (though I'm not entirely sure what Hugh's position on that is).

> I don't want much to hear that we should do X better if only we had Y, because that's not being constructive (unless you provide a PR for Y of course :) )

How am I supposed to provide a PR for something I don't want to be added?

**Comment by ousado** [comment]
@ncannasse 
Regarding d) How does an API like `haxe.MainLoop.hasEvents()`, `haxe.MainLoop.add(..)`, `haxe.MainLoop.addThread(..)`, etc. which, correct me if I'm wrong, is the current implementation, allow for having several event loops? Will it magically choose the one the user has in mind? 

Regarding e) the performance critical issues directly follow from having only one central event loop. Many applications can be designed so that there's no need for locks at all, and certainly not for simply adding an event, regardless of the number of threads. The current API enforces locks for all but the purely single-threaded case.

What I'm talking about is having e.g. several threads each with an event loop completely independent from the other threads. If I'm not mistaken I remember an instance where you yourself used that design in Tora (isn't that right?). The proposed API doesn't permit that. That's simply not acceptable.

**Comment by hughsando** [comment]
ousado, You might be reading too much into the scope here.
Firstly, it is all about "the main thread", specifically as it applies to either:
1) interacting with windowing frameworks, or
2) preventing the command-line program from exiting with unprocessed events (timer, thread, async io)

This is to solve these two very real problems.  Solving efficient multi-thread communication is out of the scope of this change, and this does not dictate how you might choose to implement this.  It only gives you a standard way of preventing command line apps from exiting and running stuff on the main thread inside someone else's windowing framework.

This code needs the be in the same repo as haxe.Timer and haxe.http and probably cpp.vm.Thread.  And since these are in the std lib, this should be too.

Currently we get people who start a timer on the cpp target, but the program exits before the timer executes, and then I get a bug report because this is not how it works in js/flash.  Similarly, not waiting for threads to finish surprises some people and I get bug reports.  As a framework developer, I would really like to use a haxe.http async load by supporting 2 simple functions.

If you have more complicated needs, you can side-step the whole thing.  But I guess you would need to think about how you will let a developer call haxe.Timer to run something later on the main thread, as they would expect to with all the targets with a multi-platform  toolkit.

**Comment by waneck** [comment]
After giving some thought about it, and talking to Nicolas and seeing Hugh's comments about it, I understand that the use case I have in mind (mostly asynchronous IO) is not the point of this, and that it won't help there. I'd ask though, if we could please include a non-static API so that we could have other kinds of Loops that are not "THE" main loop, and allow standard library APIs (e.g. Timer and Http) to optionally select which loop they will run on. I think this change will allow this to not fuddle with the async IO use case.
Otherwise, any async IO implementation will have to override Http and other types anyway, which in my humble opinion defeat the purpose of this.

**Comment by hughsando** [comment]
Maybe another way of thinking of it is that its api is really just "runOnMainThread(f)" plus "runTimerOnMainThread(t,f)" as far as a non-framework developer is concerned.
So you could have
http.requestAsync(url, onComplete),  and onComplete would allow ```function() MainLoop.runOnMainThread( function() doSuff() )``` - No explicit mention of MainLoop in the async api.
Or maybe ```http.requestAsync(url, onComplete, ?where:(Void->Void)->Void)```
Here, "where" could default to "MainLoop.runOnMainThread", you could provide "myLoop.run" if you wanted, or just ```function(f) f()``` which could maybe be "MainLoop.runImmediately".

I guess I am saying that since the api is really only 1 function, closures can work as well as or better than instances.

As for timer, I think it is a bit different. The key function is ```delay (f:Void ‑> Void, time_ms:Int):Timer```, which is static.  Rather than providing a "where" here you can add the function to your loop object:  ```myLoop.delay (f:Void ‑> Void, time_ms:Int):Timer```.  You can of course do this today.  I say this mainly because the implementation of adding a "where" here is not immediately obvious.
If Timer.delay were made into a "dynamic" function, frameworks would be able to override this very easily, although I do not see the use case here, since overriding the main loop gets you a little mode flexibility.

In summary - MainLoop is only for the main thread.  Async apis should allow callbacks on things other than the main loop, but make it easy for it to be on the main loop.

**Comment by ousado** [comment]
My concern is this:
- people will start writing code using this API
- people wish to have their code, some async IO or whatever, run on one or more parallel event loops (and that's not an uncommon or rare use-case at all) 
- people will notice that's impossible with haxe.MainLoop
- people will realize that they will have to rewrite their code 
- if haxe.MainLoop is the only event loop available on their platform at the time, then they will realize they'll even have to *write* the event loop implementation - possibly by adopting MainLoop just enough so it's not a global thing anymore.. 

That's such a bad idea, so absolutely unnecessary and so easy to avoid, I really don't know where to start.

**Comment by bjitivo** [comment]
Can someone explain to me what it even means to run code in "multiple parallel event loops" in a single threaded program?  A single threaded program can one exactly one stream of code at a time.  That stream of code is managed by a single event loop in the most basic proposal for event loops.  If you have multiple threads (i.e. the ability for the runtime to run more than one stream of code concurrently), then each of those can run their own event loop if desired.  What would it mean for a single thread of execution to have "multiple main loops"?  When would each one run?  And what would be the advantage over collapsing everything into one main loop since that thread can only one run function at a time anyway?

**Comment by bjitivo** [comment]
I guess I'm asking for an example to back up this claim:

"people wish to have their code, some async IO or whatever, run on one or more parallel event loops (and that's not an uncommon or rare use-case at all)"

In my experience, event driven programs do not manage multiple main loops.  I would love to see an example program where more than one main loop made sense in a way that wasn't just as easily accomplished with a single main loop.

**Comment by jdonaldson** [comment]
I think it's also worth considering error catching behavior.  Async and a while(true) loop are going to behave differently.

**Comment by montibbalt** [comment]
+1 on keeping it as a haxelib for several reasons:
- I don't want my developers to use this even accidentally and relying on DCE for that is pretty weak
- It seems extremely silly to add something like this directly in haxe right now when the RC milestone is already past due, we're not even going to attempt to do it right the first time (people in here are still arguing on what "doing it right" even means), and a haxelib gives you infinitely more flexibility on developing/releasing/updating it

**Comment by ousado** [comment]
@bjitivo 
> If you have multiple threads (i.e. the ability for the runtime to run more than one stream of code concurrently), then each of those can run their own event loop if desired.

Exactly - they **should** be able to do that - but with the currently proposed implementation they can't, because there's one global MainLoop only, which is exactly my point.
> I guess I'm asking for an example to back up this claim:

> >"people wish to have their code, some async IO or whatever, run on one or more parallel event loops (and that's not an uncommon or rare use-case at all)"

Glad you're asking - I think I have a few examples handy, 
- zimg.com - "A lightweight and high performance image storage and processing system.", open source, so here's exactly where each thread in the http processor is initialized with its own event loop:
https://github.com/buaazp/zimg/blob/master/src/main.c#L386 

- RProxy:  "RProxy is a reverse proxy server written with performance and scale in mind." - also open source (just not updated on github since the author and maintainer of the open source version left the company, as far as I'm aware)
https://github.com/mandiant/RProxy/blob/master/src/rproxy.c#L1454
- Erlang
every ERS that runs on a multi-core CPU and is configured accordingly binds schedulers to OS threads in a configurable way - which essentially is the one event-loop per thread model with some sugar on top: http://erlang.org/doc/man/erl.html#+sbt

- Tora, an example from Haxe-land, also uses one event loop per thread:
https://github.com/HaxeFoundation/tora/blob/master/src/Tora.hx#L904

And there are many more - it's not like it doesn't make sense to have several threads with independent event loops that essentially do the same work and do (or even don't, but that's more debatable) require some shared, perhaps immutable state, and are demanding CPU-wise, like e.g. the SSL/TLS handling in RProxy.

**Comment by bjitivo** [comment]
I think it's reasonable for the compiler to emit a preamble that runs before the main() function that initializes an event handling loop for the thread that is about to run main(), which can thus be called the "main loop", and then continue to run the main loop after the main() function as long as there continues to be callbacks scheduled on the main loop.

I think it's also reasonable for any thread to be able to create an event loop and run it if desired.

Guaranteeing that there is a thing called the "main loop" that can have work scheduled on it allows a lowest common denominator of event loop functionality, where all haxe libs and applications can at least assume that there is one event loop onto which work can be scheduled; libraries can advertise that they always schedule work on that loop, or if they want to be more sophisticated in cases where threading performance is paramount, can also take the event loop onto which to schedule work, allowing the application to select which threads will do which work.

I would expect 99% of applications to function perfectly well without ever having to create another event loop; those specialized programs which (typically without justification beyond premature optimization :) really require scheduling work onto multiple event loops will still be able to do so.

**Comment by ousado** [comment]
Oh, and I forgot, another reason to want to do that, at least according to quite many questions on various mailing lists, especially for inexperienced users, is when they realize they have some longer-blocking-than-expected operations somewhere their main loop, and want to move them out of the way without much hassle.

**Comment by bjitivo** [comment]
They'll only get to do that on platforms with multiple threads.  And turning users not sophisticated enough to understand how to break long running work up into pieces so as not to clog up the main loop loose on threads is just a very, very bad idea in my experience :)

**Comment by ousado** [comment]
yes, I too tend to encourage them to rewrite these dreaded database drivers that all too often only come in synchronous versions into non-blocking code, but somehow they rarely listen (or never show up again) :)

**Comment by ncannasse** [comment]
Having several loops does make only sense in a multithreaded application, which requires very custom synchronization primitives, it's entirely outside of the scope of this library to provide an abstraction over this. Given that it was the main comment since my last call-for-feedback, I've merged my proposal.

**Comment by markknol** [comment]
Would it be nice if `MainLoop.add` would be `f:Float->Void` as argument, where passes the delta time? 
```haxe
MainLoop.add(function(deltaTime:Float) {
      
});
```
